warp10/src/main/java/io/warp10/script/binary/SUB.java

```diff
@@ -171,19 +164,31 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
       if (!(gts.getType() == TYPE.LONG || gts.getType() == TYPE.DOUBLE)) {
         throw new WarpScriptException(typeCheckErrorMsg);
       }
-      
-      double op = op1gts ? ((Number) op2).doubleValue() : ((Number) op1).doubleValue();
-      
-      for (int i = 0; i < n; i++) {
-       double value;
-       if (op1gts) {
-         value = ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue() - op;
-       } else {
-         value = op - ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue();
-       }
-       GTSHelper.setValue(result, GTSHelper.tickAtIndex(gts, i), GTSHelper.locationAtIndex(gts, i), GTSHelper.elevationAtIndex(gts, i), value, false);
-      }      
 
+
+      Number op = op1gts ? (Number) op2 : (Number) op1;
+
+      if (op instanceof Double || gts.getType() == TYPE.DOUBLE) {
+        for (int i = 0; i < n; i++) {
+          double value;
+          if (op1gts) {
+            value = ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue() - op.doubleValue();
+          } else {
+            value = op.doubleValue() - ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue();
+          }
+          GTSHelper.setValue(result, GTSHelper.tickAtIndex(gts, i), GTSHelper.locationAtIndex(gts, i), GTSHelper.elevationAtIndex(gts, i), value, false);
+        }
+      } else {
+        for (int i = 0; i < n; i++) {
+          long value;
+          if (op1gts) {
+            value = ((Number) GTSHelper.valueAtIndex(gts, i)).longValue() - op.longValue();
```

Comment:
ditto

Reply:
Same idea as previous comment, this time because if op is not a Double, it may be a Long or an AtomicLong.

Reply:
get it out of the loop!